fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,109 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(todoList => {
+        this.setState({
+          todos: todoList
+        });
+      });
+  }
+
+  addTodo(newTodo) {
+    fetch('/api/todos', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(newTodo) })
+      .then(response => response.json())
+      .then(added => {
+        this.setState(prevState => ({
+          todos: prevState.todos.concat(added)
+        }));
+      });
+
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+  }
+
+  toggleCompleted(todoId) {
+    const copyArr = this.state.todos;
+    for (var i = 0; i < this.state.todos.length; i++) {
+      const currentTask = this.state.todos[i];
+      if (todoId === currentTask.todoId) {
+        const copyObj = currentTask;
+        copyObj.isCompleted = !currentTask.isCompleted;
```

Comment:
This is modifying the original object, which we want to avoid. We don't need to send the whole object anyway, check the api docs again